Denote dungeon levels as cleared

ikonomov opened the pull request:
The thumbs up for the vote-for-features thread makes it seem like a popular feature.  The idea is to give some indication when a level is cleared by having all monsters on that level killed: `Level: 2 (Cleared)`

ikonomov committed:
Update automap.cpp

ikonomov committed:
Update automap.cpp

Corrected the ActiveMonsterCount check based on the poisoned water quest.  I can't figure out why it needs to be 4 when all monsters are cleared, but it's not 0.

DakkJaniels commented:
Shouldn't it be less than 4, for single player too?

kphoenix137 commented:
Should deduplicate. You have the same blocks of code 3 times.

julealgon commented:
Can you update the description with what this constitutes, since we don't have an issue detailing the idea? What does it mean to "clear" a level?

ikonomov commented:
> Should deduplicate. You have the same blocks of code 3 times. Also does this account for monsters in barrels?

It should include skeletons in barrels and sarcophagi.

ikonomov commented:
> Shouldn't it be less than 5, for single player too?

The poisoned water quest is the same for both SP and MP, so it shouldn't be any different.  In SP monsters can be spawned by Leoric, but since he's the one spawning them it shouldn't be a problem.

DakkJaniels commented:
> > Shouldn't it be less than 5, for single player too?
> 
> The poisoned water quest is the same for both SP and MP, so it shouldn't be any different. In SP monsters can be spawned by Leoric, but since there is at least one monster active when they are being spawned it shouldn't be a problem.

What does the poisoned water quest have to do with this? Isn't it four because of golems? Single player doesn't spawn four golems, right?

ikonomov commented:
https://www.youtube.com/watch?v=NKbB2Np9xnA
These are the sounds that are played when a level is cleared in Skyrim.  It kind of inspires a feeling of accomplishment.  I've found some possible sounds that we could consider playing when a level is cleared:

https://github.com/diasurgical/devilutionX/assets/5624187/adfa5310-f397-4867-84ec-c5ef4a44c89b


https://github.com/diasurgical/devilutionX/assets/5624187/d6d0a28b-fb88-4799-bfc7-101985ac9098


https://github.com/diasurgical/devilutionX/assets/5624187/e992cd9e-b689-41a5-a21a-1a7407ed789e


https://github.com/diasurgical/devilutionX/assets/5624187/4b4189fa-93be-4ff2-94de-6f2d1f53142b


https://github.com/diasurgical/devilutionX/assets/5624187/78ba8e1f-edb2-4149-9cd8-d6665cfa2d7e

ikonomov commented:
> > > Shouldn't it be less than 5, for single player too?
> > 
> > 
> > The poisoned water quest is the same for both SP and MP, so it shouldn't be any different. In SP monsters can be spawned by Leoric, but since there is at least one monster active when they are being spawned it shouldn't be a problem.
> 
> What does the poisoned water quest have to do with this? Isn't it four because of golems? Single player doesn't spawn four golems, right?

I think it is golems, but the check for the poisoned water quest is checking for 4 regardless of whether it's SP or MP.  Maybe in SP it's also 4 golems even if only 1 is possible.

StephenCWills commented:
> Single player doesn't spawn four golems, right?

I'm pretty sure it does.

https://github.com/diasurgical/devilutionX/blob/8bcf520263b30493cf3924c0fe5a7c28560617d6/Source/monster.cpp#L3467-L3473

I'm guessing on setlevels they're probably included in the `.dun` file.

FitzRoyX commented:
You get way more exp and stuff per minute limiting how much you backtrack across already explored areas. The game is also scarier when you don't know how many enemies are left. As soon as you tell people there are a few enemies left, every unopened door or fogged area becomes safe.

kphoenix137 commented:
> You get way more exp and stuff per minute limiting how much you backtrack across already explored areas. The game is also scarier when you don't know how many enemies are left. As soon as you tell people there are a few enemies left, every unopened door or fogged area becomes safe.

I have to agree with this. This type of feature seems more ideal for an IronMan mod as well, where this kind of information is necessary. The only place where this information helps the player at all is in a level where you're required to kill all the monsters, such as Poisoned Water Supply, however, that levels tells you that you cleared the level already using a palette change and a sound.

julealgon commented:
Not only that, but the concept of "cleared" itself is subjective. Some people consider an area "cleared" if they went through all the floors in it. Others need to "open all chests", others also "destroy all barrels", "activate all shrines", "discover 100% of the map", "do all quests" etc etc.

That's why I asked for a small clarification in the description.

Overall, this feels very "artificial" to be put in the vanilla game, IMHO. I'd rather make sure the game exposes the details I mentioned above easily (if not already) so concepts like this can be smoothly integrated into mods, than exposing them directly like this.

ikonomov commented:
I agree with you guys, this has an affect on the gameplay.  Specifically it does not include trapped doors and chests, which could be left behind for somebody to trigger.  However, many other qol improvements do as well, such as being able to see the location of other players on the map, for example.  I think the concept of cleared as it relates to monsters only is easier to grasp, and that by itself is still useful information.  To be honest, I'm sort of undecided myself if this is a good change or not.  Lots of people voted for it from the thread, so I thought maybe it'll be a popular feature.

kphoenix137 commented:
It isn't a bad feature by any means, I just personally think it doesn't really have a place in the official build. Having the knowledge that all the monsters in the level are dead doesn't really assist the player in any way.

ikonomov commented:
I played a bit today to see how it affects gameplay.  For my style of play at least it definitely does affect gameplay.  I can only imagine that in time once I get used to it I'll become even more the case.  All additions should be qol improvements and I'm not really sure this can be easily classified into that category unless we know if the original developers would see this as such.  Because of this I think it is perhaps better to leave it out of the official build.

The other argument is also valid, if somebody like KP doesn't see his gameplay style affected whatsoever by the change then the change has no purpose.  I guess the only argument for implementing this change is if people see it somewhere in the middle, making some improvement, yet not change too much to be in conflict with the original gameplay design.  We only have the four of us to judge, but it seems like this middle ground where this would be a good idea is pretty narrow.

chasedjones88 commented on Source/automap.cpp:
```diff
@@ -748,15 +749,24 @@ void DrawAutomapText(const Surface &out)
 	switch (leveltype) {
 	case DTYPE_NEST:
 		description = fmt::format(fmt::runtime(_("Level: Nest {:d}")), currlevel - 16);
+		if (ActiveMonsterCount == 4) {
```
If ActiveMonsterCount < 4, then re-entering a level will not denote it as cleared, correct?

Trihedraf replied:
There is always 4 when clear because they are reserved slots for golems

chasedjones88 replied:
👍

chasedjones88 replied:
might be a good thing to keep as a const then, because otherwise its a magic number

kphoenix137 replied:
We should have some method of interacting with `ActiveMonsterCount` in a way that prevents the need for doing these comparisons with `4` due to the Golems.

StephenCWills commented:
I think he's asking why you closed it.

ikonomov committed:
Merge branch 'master' into level-cleared

ikonomov committed:
Update automap.cpp

added the cleared description for the quest levels

ikonomov committed:
Update automap.cpp

ikonomov commented:
I just found the "Restore branch" button XD  Also rebased and added the description for quest levels.

Edit: Sorry for the extra commit, I don't know why but the web editor adds an extra line at the end.

kphoenix137 commented:
Given that the Lua platform has been under development and ideally most of DevilutionX's QoL features should (or will) be migrated to Lua, perhaps this feature should become a Lua mod as well?

kphoenix137 commented on Source/automap.cpp:
```diff
@@ -1438,23 +1439,37 @@ void DrawAutomapText(const Surface &out)
 		description = fmt::format(fmt::runtime(_("Level: Nest {:d}")), currlevel - 16);
+		if (ActiveMonsterCount == 4) {
+			description.append(std::string(_(" (Cleared)")));
+		}
```
```suggestion
```

kphoenix137 replied:
```suggestion
```

kphoenix137 commented on Source/automap.cpp:
```diff
@@ -1438,23 +1439,37 @@ void DrawAutomapText(const Surface &out)
 		description = fmt::format(fmt::runtime(_("Level: Crypt {:d}")), currlevel - 20);
+		if (ActiveMonsterCount == 4) {
+			description.append(std::string(_(" (Cleared)")));
+		}
```
```suggestion
```

kphoenix137 commented on Source/automap.cpp:
```diff
@@ -1438,23 +1439,37 @@ void DrawAutomapText(const Surface &out)
 		description = fmt::format(fmt::runtime(_("Level: {:d}")), currlevel);
+		if (ActiveMonsterCount == 4) {
+			description.append(std::string(_(" (Cleared)")));
+		}
```
```suggestion
```

kphoenix137 replied:
```suggestion
		if (setlevel) {
			description = std::string(_(QuestLevelNames[setlvlnum]));
		} else {
			description = fmt::format(fmt::runtime(_("Level: {:d}")), currlevel);
		}
```

kphoenix137 commented on Source/automap.cpp:
```diff
@@ -1438,23 +1439,37 @@ void DrawAutomapText(const Surface &out)
+		}
+		DrawString(out, description, linePosition);
 		return;
 	}
```
```suggestion
```

kphoenix137 commented on Source/automap.cpp:
```diff
@@ -1438,23 +1439,37 @@ void DrawAutomapText(const Surface &out)
+		}
 		break;
 	}
 
```
```suggestion

	if (ActiveMonsterCount == MAX_PLRS) {
		description.append(std::string(_(" (Cleared)")));
	}
	
```

AJenbo commented:
Indeed this would make perfect sens as a Lua event and, everything else for this is already possible from Lua and it could probably be a useful event for other mods, like giving a reward for clearing a level.

ikonomov commented:
Agreed.  Although useful and nice to have, it does have gameplay consequences which are definitely better left as optional components, especially now that we have Lua.

AJenbo commented:
It also just provides a nice generic way to add toggles for everything 🙂
